### ISSUE TITLE
Support Rancher Agent and Server on a development MAC with Docker Beta?

### ISSUE THREAD

**Issue by virtuman** [rancher/rancher#5356]
Is it possible to install Rancher server and agent on a development mac machine running Docker Beta Mac ?

I saw the video where presented was talking about deploying Jenkins via rancher on development machines, but cannot figure out how to get the stack up and running.

Successfully installed Rancher Server on the local machine, but cannot add itself as a docker host to the environment, getting various exceptions, ie:

```
sudo docker run -d --privileged -v /private/var/run/docker.sock:/var/run/docker.sock -v ~/docker/rancher-agent:/var/lib/rancher rancher/agent:v1.0.2 http://localhost:8080/v1/scripts/A6528FCF080093A1B9FF:1467928800000:Ovv5dGi2jOKYgAri2cctVGzwQew
```

will result in:

> $ docker logs -f mad_brown
> Cannot connect to the Docker daemon. Is 'docker -d' running on this host?

any other thoughts on how to i can get agent installed on docker for mac?

I'm also assuming that http://localhost:8080 isn't going to work, since the agent container will literally attempt to contact itself from within agent container instead of port 8080 on the host machine?

thank you.

**Comment by mkulumadzi** [comment]
@virtuman I'm fairly new to docker and brand new to rancher, but I was able to get _somewhere_ by putting in my Mac's local ip (ie 192.168.0.x) in lieu of localhost above. Doing this allowed the agent to start and find the host.

I say _somewhere_ because I'm not able to connect to any of the services I'm adding... but one thing at a time. fwiw I'm on Docker version 1.12.0-rc3, build 91e29e8.

**Comment by vincent99** [comment]
The IP of the Docker for Mac VM is (AFAIK) always `192.168.64.1`, so that is what you would want to put in for the host registration URL. And as the agent IP if you try to register the VM as a host too.

Https://github.com/rancher/10acre-ranch contains a script that will start a rancher/server and add 3 host VMs to it with one call (3 separate ones, not using the Docker for Mac one as a host).

**Comment by virtuman** [comment]
But I don't use VMs, question about direct Docker of Mac.
On my network attached computers it's totally fine adding the network IP address of the server, but on laptops that connect to random wifi and sometime LAN it doesn't really have an IP other than 127.0.0.1 until I connect to a Wi Fi, and that will change if I re-connect to another wi-fi, and won't work if i'm not on a network at all.

So far, I was only creative enough to use either statically assigned IP address (but ipv6 kicks in by default and messes things up), or start up dnsmasq + create virtual hosts + pass it to the agent container when spinning it up (but neither of these seem natural and i'm sure there's a much better way of doing it, if I recall it correctly, Shannon used his own laptop during the presentation as well): 
https://docs.docker.com/engine/reference/run/#network-settings
`--add-host=""      : Add a line to /etc/hosts (host:IP)`

any better solutions for this?

Thank you.

**Comment by JustinPullen** [comment]
This would really accelerate our local development and local orchestration.

**Comment by vincent99** [comment]
@virtuman Docker for Mac has some fancy tricks to pretend it's not, but at the end of the day it's still just a VM funning Docker in Linux (running under the native Hypervisor.framework instead of VirtualBox).

In the earlier betas it had an IP and could be directly addressed, but this is no longer the case.  As you said this is a problem because there is no fixed IP for the agent to refer to to find the server.

**Comment by virtuman** [comment]
Maybe i'm not doing something correctly, but pinging 192.168.42.1 seems to not find that host:

```
$ ping 192.168.42.1
PING 192.168.42.1 (192.168.42.1): 56 data bytes
Request timeout for icmp_seq 0
```

**Comment by vincent99** [comment]
Yeah, later betas made it harder to access the VM when the default networking model changed.  I haven't looked lately but it is probably not possible to use the same approach anymore

**Comment by nunofgs** [comment]
I took some time to investigate this and got it working. The Docker VM can access the host with ip `192.168.65.2`, so that's not a problem (don't forget to set this URL in your rancher _Admin -> Settings_).

I first checked the docker logs on the `rancher-agent` container and found:

```
time="2016-09-02T12:56:21Z" level="info" msg="Watching state directory: /var/lib/rancher/state/containers"
time="2016-09-02T12:56:21Z" level="fatal" msg="json: cannot unmarshal bool into Go value of type string"
```

I then exec'ed into the rancher-agent container and tailed the `/var/log/rancher/agent.log`. It was full of:

````
2016-09-02 13:04:17,434 INFO process-manager [140171185050640] [process_manager.py:68] Launched ['host-api', '-cadvisor-url', 'http://127.0.0.1:9344', '-logtostderr=true', '-ip', '0.0.0.0', '-port', '9345', '-auth=true', '-host-uuid', '03822dfd-90a2-4d9e-8e98-bf92637a263e', '-public-key', '/var/lib/cattle/etc/cattle/api.crt', '-cattle-url', 'http://192.168.65.2:8080/v1', '-cattle-state-dir', '/var/lib/rancher/state/containers'] as pid 5124
2016-09-02 13:04:44,697 INFO process-manager [140171185050640] [process_manager.py:79] Process 5124 is dead, return code 0
```

I then installed strace and tried to figure out what was happening:

```
strace -s 1000 -f ./host-api -cadvisor-url http://127.0.0.1:9344 -logtostderr=true -ip 0.0.0.0 -port 9345 -auth=true -host-uuid 03822dfd-90a2-4d9e-8e98-bf92637a263e -public-key /var/lib/cattle/etc/cattle/api.crt -cattle-url http://192.168.65.2:8080/v1 -cattle-state-dir /var/lib/rancher/state/containers
```

This is what I saw just before the crash:

```
[pid  4470] <... read resumed> "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nServer: Docker/1.12.1 (linux)\r\nDate: Mon, 05 Sep 2016 19:41:21 GMT\r\nContent-Length: 213\r\n\r\n{\"Version\":\"1.12.1\",\"ApiVersion\":\"1.24\",\"GitCommit\":\"23cf638\",\"GoVersion\":\"go1.6.3\",\"Os\":\"linux\",\"Arch\":\"amd64\",\"KernelVersion\":\"4.4.19-moby\",\"Experimental\":true,\"BuildTime\":\"2016-08-18T17:32:24.504694950+00:00\"}\n", 4096) = 353
[pid  4471] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  4470] epoll_ctl(4, EPOLL_CTL_DEL, 3, {0, {u32=0, u64=0}} <unfinished ...>
[pid  4471] <... select resumed> )      = 0 (Timeout)
[pid  4470] <... epoll_ctl resumed> )   = 0
[pid  4471] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  4470] close(3 <unfinished ...>
[pid  4471] <... select resumed> )      = 0 (Timeout)
[pid  4470] <... close resumed> )       = 0
[pid  4471] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  4470] write(1, "\33[31mFATA\33[0m[0000] json: cannot unmarshal bool into Go value of type string \n", 78 <unfinished ...>
[pid  4471] <... select resumed> )      = 0 (Timeout)
FATA[0000] json: cannot unmarshal bool into Go value of type string
```

So it looks like the `host-api` process is quitting because of a bool value in the docker version response. The only thing that fits the bill is: `\"Experimental\":true`.

Armed with that knowledge, I downgraded my beta Docker.app to stable by downloading the dmg [here](https://docs.docker.com/docker-for-mac/) and everything works!

tl;dr: `host-api` has a bug that crashes on a docker version response in the beta version. Download the stable version instead as a workaround.

**Comment by jk** [comment]
> Armed with that knowledge, I downgraded my beta Docker.app to stable by downloading the dmg here and everything works!

@nunofgs So basically you installed Docker for Mac (Stable Channel)? That does not works for me (stable is now an Docker v1.12.5)

**Comment by jaredallard** [comment]
We really need this!

**Comment by vincent99** [comment]
This has only gotten less practical with CNI in 1.2, mapping ports from the moby VM (Docker for Mac) doesn't expose them onto the Mac's localhost.  Docker for Mac and their tricks to try and pretend it's native instead of a Linux VM just don't work well for Rancher's use-case, and I don't see that getting better.

We may eventually make something like "Rancher for Mac" that packages things up to run RancherOS VMs on xhyve in a way that works nicely.  In the meantime you could try the little script I use for running it on my mac when not near an internet connection: https://github.com/rancher/10acre-ranch/#docker-machine-on-macos

**Comment by jaredallard** [comment]
@vincent99 Yeah, I can see the reasoning behind it. I've gotten something like 10acre-ranch working to setup nice developer environments to mimic prod for us, and docker-machine has saved the day. At least there is an easy*ish* fix!
